Data/models/DataController.swift

```diff
@@ -72,10 +78,33 @@ public class DataController: NSObject {
         return backgroundContext
     }
     
+    public static func performBackgroundTask(save: Bool = true, _ code: @escaping (NSManagedObjectContext) -> Void) {
```

Comment:
I think if we made everything pass in a context, and if we know these are all writish operations, assert on the `viewContext` being passed in. This would be a great stopping point to catch any main thread writes. It would alos allow us to simplify all of the above logic I believe (see those code snippets about the `if context = context {}` code.